fix(source-mongodb-v2): Use Debezium's change stream for resume token validation

ed-kyu opened the pull request:
## What
<!--
* Describe what the change is solving. Link all GitHub issues related to this change.
-->
Fixes false-positive invalid resume token detection that caused unnecessary full refreshes in MongoDB CDC sync. (with "Invalid CDC position behavior" set to "Re-sync data")

Related issue: Resume tokens saved by Debezium may reference events from any collection (including unsynced collections), but PR #37557 introduced filtered change stream validation that cannot find these tokens, incorrectly marking them as invalid.

### My Problem Description

Environment:
- MongoDB Atlas cluster with 72+ hours oplog retention
- Another high-activity collection exists in the same database but was not included in the sync

Symptom:
- Despite sufficient oplog retention (72+ hours) and short sync intervals (10 minutes), sync randomly triggered a full refresh with the error: "Invalid resume token .... Saved offset is not valid. Airbyte will trigger a full refresh."

## Related Issues
This PR may also fix the following related issues:
- #70222

## How
<!--
* Describe how code changes achieve the solution.
-->
Modified MongoDbDebeziumStateUtil.isValidResumeToken() to use a fallback validation strategy:

1. First: Try cluster-level mongoClient.watch(BsonDocument.class) validation without filters - this accurately validates tokens from any collection in the oplog
2. Fallback: If unauthorized (MongoDB error code 13), fall back to filtered change stream validation for environments with restricted permissions
Refactored the filtered validation logic into a separate validateWithFilter() method for clarity.

## Review guide
<!--
1. `x.py`
2. `y.py`
-->
1. MongoDbDebeziumStateUtil.java - Main logic change in isValidResumeToken() method (lines 95-130) and new validateWithFilter() method (lines 132-182)
2. MongoDbDebeziumStateUtilTest.java - Updated existing tests to mock cluster-level watch, added new testIsResumeTokenValidWithFallbackToFilteredValidation() test

## User Impact
<!--
* What is the end result perceived by the user?
* If there are negative side effects, please list them. 
-->
- Positive: Users will no longer experience unexpected full refreshes when Debezium saves resume tokens from unsynced collections (e.g., high-activity collections that are not part of the sync configuration)
- No negative side effects: The change maintains backward compatibility - filtered validation is still used as a fallback for users with restricted MongoDB permissions

## Can this PR be safely reverted and rolled back?
<!--
* If unsure, leave it blank.
-->
- [x] YES 💚
- [ ] NO ❌

ed-kyu committed:
fix: Use cluster-level change stream for resume token validation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=71049) <br/>All committers have signed the CLA.

ed-kyu commented:
Hi @rodireich! 👋 Just checking in on this PR. Let me know if any changes are needed or if there's anything I can clarify. Thanks!

rodireich commented:
Thanks for pinging me @ed-kyu ! I missed this PR

Ignore for a moment the permission, let's assume we're doing it cluster level and the user has the `readAnyDatabase` privilege:
Your change is getting a change stream using ``` mongoClient.watch(BsonDocument.class) ```, which is how we used to do it before the [change](https://github.com/airbytehq/airbyte/pull/37557/files#diff-a9e8845c74dcef64ca36c8424f841a693af0771702c9721a202c805fc2b960caL114) in #37557.
The problem with that was that we we start CDC session, debezium - our CDC library - is first building a pipeline of the included streams and then is generates a change stream using ``` mongoClient.watch(pipeline, BsonDocument.class) ```, how we changed our code to be.

I suspect that your change of reverting back to the initial `watch()` call without pipeline would generate false positives, that is a token we identified as valid is failing to resume since it points to a stream not included in the sync. 
This was the original problem that led to #37557

ed-kyu commented:
Thanks for reviewing, @rodireich!
I see that #37557 was originally created to fix permission issues ('user not authorized' errors), not false positives.

Regarding the false positive concern: my understanding is that `resumeAfter(token)` starts reading from the **position after** the token, regardless of whether that specific event matches the pipeline filter. The pipeline then filters subsequent events.

So even if the saved token points to an event from an unsynced collection:
1. Validation: passes (position exists in oplog)
2. Actual CDC: starts after that position → pipeline filters to synced collections only

Could you clarify if there was an actual case where a token was validated successfully but Debezium failed to resume? I want to make sure I'm not missing something.

ed-kyu commented:
Thanks for reviewing, @rodireich!

I understand your concern. You're right that my PR creates an inconsistency:
- Token creation: uses pipeline
- Validation: cluster-level (no pipeline) 
- Debezium usage: uses pipeline

This violates MongoDB's recommendation to use the same pipeline for resume operations.

However, I want to highlight the trade-off:

**Current problem (pipeline-only validation):**
False negatives occur when the resume token points to an event from an unsynced collection. This causes unnecessary full refreshes even when the token is still valid in oplog - which is the issue I encountered.

**My PR approach:**
Solves false negatives, but technically outside MongoDB's documented contract.

**Questions:**
1. Was there an actual case where cluster-level validation passed but Debezium failed to resume? Or is this a theoretical concern based on MongoDB's warning?
2. In my testing: cluster-level validation success → Debezium resume succeeded.
3. Do you have suggestions for an alternative approach that avoids both issues?

I'm open to other solutions if there's a better way to handle this.

ed-kyu commented:
Hi @rodireich,

After deeper analysis, I’d like to share my findings and discuss possible approaches.

## The Root Cause

The issue stems from a limitation of pipeline-based validation. Even when token generation and validation use the same pipeline, a resume token fundamentally represents a position in the oplog rather than being strictly tied to a specific synced collection.

When getResumeToken() is called, it captures the current oplog position, which may correspond to an event from any collection, including unsynced ones. During pipeline validation, this can cause validation to fail if the token points to an event that is later filtered out by the pipeline.

In such cases, actual CDC is expected to succeed, since it resumes from that oplog position and then applies the pipeline to filter and emit only the synced collections.

## Options I’ve Considered
### Option 1: Current PR (Cluster-level validation with fallback)
Cluster-level validation first, with a fallback to pipeline-based validation on permission errors (code 13).
- Pros: Eliminates known false negatives
- Cons: May conflict with MongoDB’s recommendation to resume using the same pipeline and options

### Option 2: Remove validation entirely (Debezium-style)
Attempt CDC resume directly and handle failures at runtime.
- Pros: Avoids validation mismatches entirely
- Cons: Potentially worse UX, as users may encounter sync failures instead of an early full refresh

### Option 3: Two-level validation (validation as a hint, not the authority)
```java
boolean pipelineValid = validateWithPipeline(...);
if (!pipelineValid) {
    boolean clusterValid = validateClusterLevel(...);
    if (clusterValid) {
        // Likely points to an unsynced namespace; attempt CDC and let runtime resume be authoritative
        LOGGER.warn("Resume token may reference an unsynced namespace");
        return true; // Proceed with CDC; actual resume determines success
    }
    // Both validations failed: treat as non-resumable
    return false;
}
```
- Pros: Pipeline validation remains the primary signal; cluster-level validation only mitigates false negatives
- Cons: Slightly more complex logic

I’d appreciate your thoughts on these options and any guidance you may have.

ed-kyu committed:
refactor: Use Debezium MongoUtils.openChangeStream() for resume token validation

ed-kyu committed:
revert: Restore mongodb-v2 documentation to original

ed-kyu commented:
@rodireich 
Done! I've updated the validation to use `MongoUtils.openChangeStream(mongoClient, taskContext)` which internally uses `ChangeStreamPipelineFactory` to build the same pipeline as Debezium's streaming phase. This ensures our token validation matches 1:1 with Debezium's actual resume logic.

Also updated the tests to mock `isValidResumeToken` since Debezium's internal classes don't work well with mocked MongoClient instances.

Let me know if you have any feedback or if anything needs to be adjusted!

rodireich commented on airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:
```diff
@@ -164,9 +165,22 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
+
+    // Create Debezium properties for validation - using same pipeline as CDC resumption
+    final var validationOffsetManager = AirbyteFileOffsetBackingStore.initializeState(cdcState.state(), Optional.empty());
+    final var validationPropertiesManager =
```
This should set the `capture.scope` correctly which will ensure that for a `deployment` scope, the new code will validate the token with `client.watch(pipeline…` while for `database` scope it will use `client.getDatabase(database).watch(pipeline…` similar to how we do today.

Looks good

rodireich committed:
Merge branch 'master' into edkyu/use-cluster-level-change-stream-for-resume-token-validation

rodireich committed:
Add version 2.0.6 release notes for MongoDB source

rodireich committed:
Update MongoDB source connector docker image tag
